module1/homework/grayscaleImages/compression.cpp

```diff
@@ -0,0 +1,67 @@
+#include "compression.hpp"
+
+#include <iostream>
+#include <string>
+
+std::vector<std::pair<uint8_t, uint8_t>> compressGrayscale(
+    std::array<std::array<uint8_t, height>, width>& uncompressed) {
+    std::vector<std::pair<uint8_t, uint8_t>> compressed;
+    compressed.reserve(height * width);
+
+    for (const auto& row : uncompressed) {
+        auto rowIterator = row.cbegin();
+        uint8_t value = *rowIterator;
+        uint8_t currentValueCounter = 1;
+        rowIterator++;
+
+        while (rowIterator != row.cend()) {
+            if (*rowIterator == value) {
+                currentValueCounter++;
+            } else {
+                compressed.push_back(std::make_pair(value, currentValueCounter));
+                value = *rowIterator;
+                currentValueCounter = 1;
+            }
+            rowIterator++;
+        }
+        // and for last value in a row:
+        compressed.push_back(std::make_pair(value, currentValueCounter));
+    }
+
+    compressed.shrink_to_fit();
+    return compressed;
+}
+
+std::array<std::array<uint8_t, height>, width> decompressGrayscale(
+    std::vector<std::pair<uint8_t, uint8_t>>& compressed) {
+    std::array<std::array<uint8_t, height>, width> uncompressed;
+    auto pixelGroup = compressed.begin();
+    uint8_t howManyPixels = pixelGroup->second;
+
+    for (auto& row : uncompressed) {
+        for (auto& pixel : row) {
+            if (howManyPixels == 0) {
+                pixelGroup++;
+                howManyPixels = pixelGroup->second;
+            }
+            pixel = pixelGroup->first;
+            howManyPixels--;
+        }
+    }
+    return uncompressed;
+}
+
+void printMap(std::array<std::array<uint8_t, height>, width>& bitmap) {
+    const std::string palette = R"( .'`^",:;Il!i><~+_-?][}{1)(|\/tfjrxnuvczXYUJCLQ0OZmwqpdbkhao*#MW&8%B@$)";
+    int level;
+    const double normalizingFactor = palette.size() / 256.0;
+    std::cout << "\n";
+
+    for (auto& row : bitmap) {
+        for (auto& pixel : row) {
+            level = pixel * normalizingFactor;
+            std::cout << palette[level];
+        }
+        std::cout << "\n";
+    }
```

Comment:
Interesting. I don't really know why you complicated it like this. You want to print values as ASCII art :)?
If yes just `use std::cout << pixel`; Because ostream iterator treat uint8_t like unsigned char and will print characters not numbers.

Reply:
I thought I had to map pixel's brightness to appropriate character - the brightest are $, @, the darkest are a space (brightness == 0), dot, apostrophe, etc.
If I did it as you suggest, I would get characters with "random" brightness. 
That's why I used that palette. One can use different palette, for example " .:-=+*#%@", and my implementation will automatically normalize brightnesses to that shorter one.

Reply:
As @ziobron suggested on discord, I pasted the output of my implementation.